### ISSUE TITLE
Unable to login or register

### ISSUE THREAD

**Issue by 20Koen02** [ajwad-shaikh/BusKaro#14]
**Describe the bug**
Can't login or create a user. Fails wirh the error: `Fatal error: Uncaught Error: Call to a member function fetch_assoc() on boolean in C:\wamp64\www\BusKaro\index.php on line 154`

**To Reproduce**

1. Start WAMP
2. Go to localhost/buskaro
3. Try to login or register
4. Notice error on the bottom of the website

**Expected behavior**
The user can log in or register

**Desktop (please complete the following information):**

- OS: Windows 10 Pro
- Browser Chrome

**Comment by madhavgoyal98** [comment]
I would like to help on this issue

**Comment by ajwad-shaikh** [comment]
Sure, @madhavgoyal98

**Comment by EuricoMazivila** [comment]
Buskaro has not yet implemented features that prevent SQL Injection attacks, but what is this about SQL Injection:
SQL Injection
Technique that malicious users use to “inject” SQL code into a lawful SQL statement using Web form fields, the URL, or other input means;
It may compromise the security of the web application;
It can enable unauthorized query, update, and delete operations to be performed on the database.

To avoid SQL Injection the following features are used:

**Avoiding SQL Injection**

- Preprocessing (validating) data entered in form (or URL) fields is recommended to prevent SQL Injection attacks;
// Validation of Populated Data in Form Fields
     function input_filter ($data) {
         // remove spaces at the beginning and
         // at the end of the string
         $data = trim ($data);

         // remove against slashes:
         // "water snake" becomes "water snake"
         $data = stripslashes ($data);
         $data = htmlspecialchars ($data);

         return $data;
     }
The PHP htmlspecialchar function, for example, removes some special characters from the string, such as quotation marks, by replacing these characters with the corresponding HTML codes;

> In addition to validating data from form fields or URLs, it is also recommended not to enter the name of PHP variables directly into the SQL string;

Instead, use the concept of prepared statements (presented below).

**Prepared Statement** is a feature that allows you to execute the same SQL statement over and over, more efficiently and more safely.
Main advantages:
• Greater security against SQL Injection attacks.
• May reduce the execution time of multiple SQL queries on the server, as their preparation can be performed only once (although the query may be executed repeatedly);
• Can reduce data traffic with the server as the SQL statement does not have to be resubmitted in full again and again;

Basic operation:
• Preparation: An SQL template is created and sent to the DBMS. However, certain values ​​are left open (called parameters). Example: INSERT INTO Customers VALUES (?,?,?)
• DBMS preprocesses the SQL statement (checks the syntax, among others), but does not execute it;
• Execution: later, the application gives values ​​to the parameters and executes the declaration;

**Can I implement these features (preprocessing / validating and prepared statements) that prevent SQL Injection attacks?**

**Comment by ajwad-shaikh** [comment]
Thanks @EuricoMazivila for the inputs. Indeed, defence against SQL Injection is crucial for the project. We can either split this issue into two or if we don't get a response from @madhavgoyal98 soon, we can have you assigned on this one so you can resolve both the issues in under #14

Let me know what you think!

Thanks again :)

**Comment by EuricoMazivila** [comment]
That's fine by me, but for me to start will depend a lot more on the response of @EuricoMazivila madhavgoyal98…
I have a question- Which of the 15 databases should I use if I start with my contributions?
Thank you so much for letting me participate…

**Comment by ajwad-shaikh** [comment]
Can @EuricoMazivila create a new issue and work under that for SQL Injection?
This issue is closed now
